### ISSUE TITLE
Error signing with Android keystore key class javax.crypto.IllegalBlockSizeException: null

### ISSUE THREAD

**Issue by crashvb** [schwabe/ics-openvpn#1289]
To make issues more manageable, I would appreciate it if you fill out the following details as applicable:

# General information
1. Android Version: 11
2. Android Vendor/Custom ROM: Google
3. Device: Pixel 5 (stock / not rooted)
4. Version of the app (version number/play store version/self-built): v0.7.21, from Play Store
 
# Description of the issue

Unable to use p12 certificates from Android Hardware Keystore. It appears the the `<key />` section is omitted from the generated configuration.

Steps to reproduce.

1. Generated a root CA, intermediate CA, server, and client certificates (8K keys).
2. The root, intermediate, and server certificates are deployed to openvpn > 2.4 on pfsense.
3. A p12 was generated containing the root, intermediate, and client certificates, and the client key.
4. Add the p12 to the Android key store. Either through "System" > "Encryption & credentials" > "Install a certificate", or using the OpenVPN for Andriod UI under "Basic" > "Install new certificate" (the result is the same either way) -- Provide the corresponding password with prompted.
5. Configure the VPN using "Android Certificate" and select the imported certificate. -- errors out as logged below.
6. (Counter test). Change the type from "Android Certificate" to "PKCS12 File" and select the original p12 that was imported, and provide the corresponding password. -- everything works fine. The generated configuration is changed to use `<pkcs12 />` instead. A single warning about using "auth-nocache" is returned (as expected) because of the password caching.

```log
021-01-17 18:34:58 VERIFY X509NAME OK: CN=router0.redacted ...
2021-01-17 18:34:58 VERIFY OK: depth=0, CN=router0.redacted ...
2021-01-17 18:34:58 Error signing with Android keystore key class javax.crypto.IllegalBlockSizeException: null
```

```config
# Config for OpenVPN 2.x
# Enables connection to GUI
management /data/user/0/de.blinkt.openvpn/cache/mgmtsocket unix
management-client
management-query-passwords
management-hold

setenv IV_GUI_VER "de.blinkt.openvpn 0.7.21" 
setenv IV_SSO openurl,crtext
setenv IV_PLAT_VER "30 11 arm64-v8a google redfin Pixel 5"
machine-readable-output
allow-recursive-routing
ifconfig-nowarn
client
verb 4
connect-retry 2 300
resolv-retry 60
dev tun
remote redacted 1194 udp
 connect-timeout  60
### From Keystore/ext auth app ####
<ca>
-----BEGIN CERTIFICATE-----
MIIJuzCCBaOgAwIBAgIIcljP6duTKkUwDQYJKoZIhvcNAQELBQAwazEeMBwGA1UE
redacted
5qLlGcBLcTPp9Lr0PrJZAzguQOa1hi4X+1u3OBIx2FxFWZDA8sQAa5vCBoSkvBE=
-----END CERTIFICATE-----

</ca>
<cert>
-----BEGIN CERTIFICATE-----
MIIKbjCCBlagAwIBAgIIQJUn10cf7zMwDQYJKoZIhvcNAQELBQAwazEeMBwGA1UE
redacted
+WMzXZVMvu9fMj5w8yH9Kg4fWZ02CTT7ph/o4x9LojwIgw==
-----END CERTIFICATE-----

</cert>
management-external-key nopadding
<tls-crypt>
#
# 2048 bit OpenVPN static key
#
-----BEGIN OpenVPN Static key V1-----
redacted
-----END OpenVPN Static key V1-----


</tls-crypt>
nobind
verify-x509-name router0.redacted name
remote-cert-tls server
data-ciphers AES-256-GCM
auth SHA256
# Use system proxy setting
management-query-proxy
# Custom configuration options
# You are on your on own here :)
compress
cipher AES-256-GCM
verb 5


```

**Comment by schwabe** [comment]
> 2021-01-17 18:34:58 Error signing with Android keystore key class javax.crypto.IllegalBlockSizeException: null

To be honest 8k RSA keys are rare. I don't know if they are actually supported by the android keystore.

**Comment by crashvb** [comment]
I was able to find a referenced to this keysize under the android built-in [keypair generator](https://android.googlesource.com/platform/frameworks/base/+/master/keystore/java/android/security/keystore/AndroidKeyStoreKeyPairGeneratorSpi.java#117); however, I don't know to which release(s) this commit was added. I suspect that it is already active / applied as the introducing [commit](https://android.googlesource.com/platform/frameworks/base/+/21a76df55cf4b956f4d34f57c7b9e694d0363f54) was submitted some time prior to 2015 ... other commits from contributing branches are dated 2013.  I'm working under the assumptions that 1) code on the master branch made it into the Android 11 release, and 2) if the built-in keypair generator can use 8K, then I should be able to import 8k chains.

Additionally, I have been researching `management-external-key nopadding`, and the management interface that is used to request signing with the private key. I re-executed the test with `verb 5` and I'm able to see the both the surrounding TLS exchange, and management calls, but I don't believe that I have isolated the individual call that is returning the null cyphertext.

I don't have experience in mobile app development, but I'd like to resolve this issue if possible. Could you point me in the right direction to be able to isolate and perform a integration test against the management interface using an 8k keypair? Everything I've been googling has been pointing me to AndroidStudio, the `<module>/src/androidTest/java/` location, and relies on emulated hardware; however, this repository only appears to have a `ics-openvpn/main/src/test/java/de/blinkt/openvpn/core/` location. What / how should I test that is in alignment with _this_ project?

**Comment by schwabe** [comment]
The project does lack unit tests since it is a spare time project and I were not always in the mood to write unit tests. Testing the hw keystore with unit tests is next to impossible anyway. 

For debugging this, you probably want to look at the `getKeyChainSignedData` function in VpnProfile as that is the one that does the actual signing. 

The commit you found is for generating keys. In this case we assume that the key is already imported into keystore. 

And just because the API can speak 8k doesn't mean that the hardware can do so as well. Take the popular YubiKeys. The smartcard API can do 8k keys but yubikeys can also only do 4k keys.

In general (not limited to Android) I found that RSA support > 4k is very spotty and limited and you are better of by usiung secp384r1 or secp521r1 if you want more security.

A good test would be if the problem can be reproduced with 4k keys or not.

**Comment by crashvb** [comment]
> The project does lack unit tests since it is a spare time project and I were not always in the mood to write unit tests. Testing the hw keystore with unit tests is next to impossible anyway.

No judgement; thank you for taking the time to share an awesome project that I've already been using for many years (might have been better if I'd started with that). I've been so accustom to using this app to remote into different VPNs, that I've overlooked that it's not built in part of Android.

> For debugging this, you probably want to look at the `getKeyChainSignedData` function in VpnProfile as that is the one that does the actual signing.

Thank you for the method name; this saves me from having to dig around to find where the magic happens.  Right now I'm leaning towards isolating the API call and the most minimum context possible to see if I can retrofit a test app and ?sideload? it onto my device.

> And just because the API can speak 8k doesn't mean that the hardware can do so as well. Take the popular YubiKeys. The smartcard API can do 8k keys but yubikeys can also only do 4k keys.
>
> A good test would be if the problem can be reproduced with 4k keys or not.

Agreed; I haven't been able to find any supporting documentation about the HSM that comes with the Pixel 5 (but I haven't been looking that hard either). I was able to use 8k keys from the hardware keystore for both WiFi authentication, and as TLS client certs through chrome to EJBCA (which I know does a great job of checking all the little details). So while I'm not 100% sure that there are no edge cases, I have high enough confidence to start with the assumption that they should work; I'll circle back around to setting up a 4k test environment if I hit a deadend.

**Comment by crashvb** [comment]
@schwabe, I got a little time to circle back around to this, and I discovered a few things ...

TLDR; The learning curve for Android is a little steep. I could use a little more help.

I did, ambitiously, attempt to clone and build the app from source to see if I could add some additional logging / stack traces ... this did not go well ... I think I'm missing just about all of the build tools.

Next I attempted to isolate what was happening by tracing the code, based on the exception that was being thrown, and the method name that you provided (In retrospect, I think you save me a day or two with that pointer).

This is what I found:

> 1. mAuthenticationType == TYPE_EXTERNAL_APP ... as that is the only option that adds `management-external-key nopadding` to the configuration.
> 
> 2. javax.crypto.IllegalBlockSizeException is caught and logged at:
>   - tlsexternalcertprovider/src/main/java/de/blinkt/externalcertprovider/ExternalCertService.java:55 ... appears to be another process? thread? package? ... how does Android do this? injection?
>   - main/src/main/java/de/blinkt/openvpn/VpnProfile.java:1219 ... from alternate branch in VpnProfile::getSignedData() -> VpnProfile::getKeyChainSignedData() -> **outside of try catch** VpnProfile::processSignJellyBeans() -> NativeUtils::rsasign() **deadend** ... assuming it needs to have the private key in memory to be able to pass it to openssl.
> 
> <Knowledge Gap>
> 
> 3. VpnProfile::getSignedData() -> VpnProfile::getExtAppSignedData() -> ExtAuthHelper::signData() -> ExternalCertificateProvider::getSignedDataWithExtra() ... jumps to AIDL (Android Interface Definition Language) ... assuming this is like a syscall to the OS, but it's defined under this apps package and googling returns no results? ... is this vendor provided, or native code, or ???
> 
> <Knowledge Gap>
> 
> 4. Starting from scratch: https://developer.android.com/training/articles/keystore#SigningAndVerifyingData
> 
> 5. References to AndroidKeyStore are at:
>   - main/src/ui/java/de/blinkt/openvpn/fragments/KeyChainSettingsFragment.kt:52 ... .kt is for Kotlin ... look kinda like groovy? **sidetracked**
> 
> 6. KeyChainSettingsFragment::setCertificate() -> ExtAuthHelper.getCertificateChain() ... around we go again ...

I _believe_ that the successful invocation of ExtAuthHelper.getCertificateChain() from within the kotlin fragment confirms that the external service and the underlying HSM are able to handle 8k keys, even if the fragment is only retrieving the metadata.

Where I could use some help is tracing beyond the getSignedDataWithExtra() protype in the AIDL file, and confirmation (or rejection) that it's implementation is part of this application or the OS.

As I'm not really interested in becoming an App developer at this time (_but I am still interested in getting this to work_), and the only language I've been able to "homebase" around for this project is java, I would like to try and continue to isolate a code fragment for testing, but instead of trying to build a whole I think I just try and [build a simple](https://raccoon.onyxbits.de/blog/run-java-app-android/) jar.

I do anticipate problems with this approach as I have a hunch that I will need a `Context` in order to directly copy the code; Do you know if a context is required for the external signing operation?

**Comment by schwabe** [comment]
> mAuthenticationType == TYPE_EXTERNAL_APP ... as that is the only option that adds management-external-key nopadding to the configuration.

You got fooled by the missing break in the switch statements:

```
            case VpnProfile.TYPE_USERPASS_KEYSTORE:
                cfg.append("auth-user-pass\n");
            case VpnProfile.TYPE_KEYSTORE:
            case VpnProfile.TYPE_EXTERNAL_APP:
```
> ExternalCertService is if you another app doing the actual signing etc. You got onto the wrong track there.

> . .kt is for Kotlin ... look kinda like groovy? sidetracked

Yes, it is Kotlin. The new default language for Android development.

> I do anticipate problems with this approach as I have a hunch that I will need a Context in order to directly copy the code; Do you know if a context is required for the external signing operation?

Yes. But Context is pretty required for almost all non-basic Android stuff.

**Comment by schwabe** [comment]
I just tried a simple 8k RSA certificate on my Pixel 5 and the result are pretty broken as well. It just hangs at the verification step. 4k RSA certificates work without problem.

(I used `openssl req -x509 -newkey rsa:8192 -keyout tyche.pem -out tyche.pem -nodes -sha256 -days 3650 -subj '/CN=tyche'` and `openssl pkcs12 -nodes -export -in tyche.pem  -inkey tyche.pem   -name "tyche" -out tyche.p12` to generate the certificate)

**Comment by schwabe** [comment]
My Nokia 3.1 test device (running Android 10) does not even import the 8k RSA in the keystore:

```
1-21 11:30:12.785   354   354 D TlcTeeKeyMaster: TEE_ImportKey [vendor/mediatek/proprietary/trustzone/trustonic/source/external/keymaster/common/410/3.0/src/tlcTeeKeymaster_if.cpp:1266]
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: params =
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_ALGORITHM: 0x00000001
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_PURPOSE: 0x00000002
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_PURPOSE: 0x00000003
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_PURPOSE: 0x00000000
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_PURPOSE: 0x00000001
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_PADDING: 0x00000001
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_PADDING: 0x00000002
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_PADDING: 0x00000004
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_PADDING: 0x00000005
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_PADDING: 0x00000003
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_DIGEST: 0x00000000
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_DIGEST: 0x00000001
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_DIGEST: 0x00000002
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_DIGEST: 0x00000003
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_DIGEST: 0x00000004
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_DIGEST: 0x00000005
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_DIGEST: 0x00000006
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_NO_AUTH_REQUIRED: true
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_ORIGINATION_EXPIRE_DATETIME: 0x7fffffffffffffff
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_USAGE_EXPIRE_DATETIME: 0x7fffffffffffffff
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: KM_TAG_ACTIVE_DATETIME: 0x0000000000000000
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster:  [vendor/mediatek/proprietary/trustzone/trustonic/source/external/keymaster/common/410/3.0/src/tlcTeeKeymaster_if.cpp:1267]
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: key_format = 0x00000001 [vendor/mediatek/proprietary/trustzone/trustonic/source/external/keymaster/common/410/3.0/src/tlcTeeKeymaster_if.cpp:1268]
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster: key_data = <data of length 4680>
01-21 11:30:12.785   354   354 D TlcTeeKeyMaster:  [vendor/mediatek/proprietary/trustzone/trustonic/source/external/keymaster/common/410/3.0/src/tlcTeeKeymaster_if.cpp:1269]
01-21 11:30:12.881  3609 21706 D SharedPreferencesImpl: Time required to fsync /data/user/0/com.google.android.gms/shared_prefs/com.google.android.gms.safetynet.xml: [<1: 0, <2: 0, <4: 0, <8: 0, <16: 3, <32: 9, <64: 3, <128: 4, <256: 3, <512: 5, <1024: 4, <2048: 0, <4096: 0, <8192: 0, <16384: 0, >=16384: 0]
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *****************************
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *** ERROR: decode_pkcs8_rsa_key:277: 'core_key_data_len >= KM_RSA_METADATA_SIZE + n_len + e_len + d_len + p_len + q_len + dp_len + dq_len + qinv_len' is false
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *** Detected in vendor/mediatek/proprietary/trustzone/trustonic/source/external/keymaster/common/410/3.0/src/km_encodings.cpp:277()
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *****************************
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *****************************
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *** ERROR: decode_key:1180: decode_pkcs8( pos, core_key_data_len, keySizeInBits, rsa_pubexp, key_encoding) == -44
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *** Detected in vendor/mediatek/proprietary/trustzone/trustonic/source/external/keymaster/common/410/3.0/src/tlcTeeKeymaster_if.cpp:1180()
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *****************************
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *****************************
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *** ERROR: TEE_ImportKey:1317: decode_key(key_format, algorithm, &keySizeInBits, &rsa_pubexp, key_data, &km_key_data, &km_key_data_len) == -44
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *** Detected in vendor/mediatek/proprietary/trustzone/trustonic/source/external/keymaster/common/410/3.0/src/tlcTeeKeymaster_if.cpp:1317()
01-21 11:30:12.920   354   354 E TlcTeeKeyMaster:   *****************************
01-21 11:30:12.920   354   354 D TlcTeeKeyMaster: TEE_ImportKey exiting with -44 [vendor/mediatek/proprietary/trustzone/trustonic/source/external/keymaster/common/410/3.0/src/tlcTeeKeymaster_if.cpp:1394]
01-21 11:30:12.920  2230 21772 I keymaster_worker: importKey failed
```

**Comment by crashvb** [comment]
> > mAuthenticationType == TYPE_EXTERNAL_APP ... as that is the only option that adds management-external-key nopadding to the configuration.
> 
> You got fooled by the missing break in the switch statements:
> 
> ```
>             case VpnProfile.TYPE_USERPASS_KEYSTORE:
>                 cfg.append("auth-user-pass\n");
>             case VpnProfile.TYPE_KEYSTORE:
>             case VpnProfile.TYPE_EXTERNAL_APP:
> ```
> 
> > ExternalCertService is if you another app doing the actual signing etc. You got onto the wrong track there.

Indeed, I was fooled; must have been sitting too close to the monitor (or I'm just rusty) ;)

> > I do anticipate problems with this approach as I have a hunch that I will need a Context in order to directly copy the code; Do you know if a context is required for the external signing operation?
> 
> Yes. But Context is pretty required for almost all non-basic Android stuff.

Thank you for saving me a few more hours here.

> I just tried a simple 8k RSA certificate on my Pixel 5 and the result are pretty broken as well. It just hangs at the verification step. 4k RSA certificates work without problem.

Ok, I yield; positive verification using the same keysize on the same device is enough for me; thank you so very much for taking the time to work this issue with me.

Based on this I'll see if I can re-issue / downgrade just the user keypair to 4K (and keep the 8K CA), or if I have to cave-in an deploy a second 4K CA and VPN instance for the pixel.

**Comment by schwabe** [comment]
user keypair with 4k should be enogh. Or not using the keystore. If you want 256 bit security strength using secp521r1 EC client certificate might also be an option.

**Comment by crashvb** [comment]
For completeness, I was able to successfully keep using the 8K root and intermediary CAs, while reducing the user keypair to 4K.

I do have a lingering curiosity, but I think that'd be better directed at support.google.com, or the XDA forums. -- Thanks again!
